### PR TITLE
[DROOLS-3993] Modified KieAssetsDropdown for better/easier reusability

### PR DESCRIPTION
**Pull request by gitgabrio**
@manstis @jomarko @karreiro 

This PR requires https://github.com/kiegroup/appformer/pull/696

I've slightly refactored original KieDropdown implementation.
My general point of view about abstract classes is
1) they are meant to be extended, so it should be possible to change the behavior, otherwise they are normal classes "masked"
2) to allow that, properties/methods must be readable/writable by child classes
3) only for very specific cases there must be an hidden/unchangeable inner state/behavior (but I do not see such case here)

For this specific one, my idea is 
1) it is GREAT to have one single elemental dropdown, to be used *everywhere* (I have at least one other case where I will use it)
2) following the above, the top-most parent must be absolutely independent of any specific constraint (e.g. IsSubmarine object)
3) the AbstractSubmarineDropDown is a specialization of the above; nevertheless, it must be possible for any actual implementation to change the behavior depending on the IsSubmarine state (for ex, my original issue was that in my scesim even if inside submarine I would need a full dropdown, not a manual input text box).

### CODE REVIEW

File: kie-wb-common-widgets/kie-wb-common-ui/src/test/java/org/kie/workbench/common/widgets/client/assets/dropdown/AbstractDropdownTest.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    protected KieAssetsDropdownItem kieAssetsDropdownItemMock;
+
+    protected KieAssetsDropdown dropdown;
+
+    protected KieAssetsDropdown.View viewMock;
+
+
+
+    protected void commonSetup() {
+        when(viewMock.getElement()).thenReturn(htmlElementMock);
```
Comment (manstis) [inline kie-wb-common-widgets/kie-wb-common-ui/src/test/java/org/kie/workbench/common/widgets/client/assets/dropdown/AbstractDropdownTest.java]: When something is essential for a sub-class test I usually add an abstract method in the base abstract class to get it. For example, I'd be tempted to do `viewMock = getMockView()` in this `commonSetup()` method and add a `protected abstract KieAssetsDropdown.View getMockView()` method. The same goes for `dropdown`.

**Comment by gitgabrio** [comment]
> Looks good to me; just a couple of comments in addition to the one inline.
> 
> 1. You need to copy `KieAssetsDropdownView.less` to a new file `SubmarineKieAssetsDropdownView.less` and change the `data-i18n-prefix` on the first line accordingly. Otherwise the `LESS` isn't picked up by Errai and applied to `SubmarineKieAssetsDropdownView.html`.

_Done. Is there a documentation for that stuff? We already have scheduled the i18n fixing for scesim following the giudlines, but I'm not sure how this is supposed exactly to work_
> 2. Could you run the code formatter over the code as there's some formatting issues.

_Done_

File: kie-wb-common-widgets/kie-wb-common-ui/src/test/java/org/kie/workbench/common/widgets/client/assets/dropdown/AbstractDropdownTest.java (hunk 1 above)
Comment (gitgabrio) [reply kie-wb-common-widgets/kie-wb-common-ui/src/test/java/org/kie/workbench/common/widgets/client/assets/dropdown/AbstractDropdownTest.java]: @manstis 

_Done _

**Comment by manstis** [comment]
@gitgabrio Re: LESS file; [this](http://docs.jboss.org/errai/latest/errai/reference/html_single/#sid-51806600_ErraiUI-ExtendedstylingwithLESS) is the only mention I can find. IIRC the `data-i18n-prefix` attribute is used (mainly) to provide a prefix for i18n however also isolates the LESS to the particular HTML file to avoid clashes across files.

**Comment by gitgabrio** [comment]
@manstis 

> @gitgabrio Re: LESS file; [this](http://docs.jboss.org/errai/latest/errai/reference/html_single/#sid-51806600_ErraiUI-ExtendedstylingwithLESS) is the only mention I can find. IIRC the `data-i18n-prefix` attribute is used (mainly) to provide a prefix for i18n however also isolates the LESS to the particular HTML file to avoid clashes across files.

_Thanks! I confused the LESS (that it is for css) with the data-i18n-prefix. This latter is actually what I would like to understand, because my understanding is there could be some kind of magic on translations, but could not find how to test/debug/make it work (so for the moment being we are using the "properties" way)_

**Comment by manstis** [comment]
@gitgabrio You will always need _properties_ files to hold the i18n text; however with HTML `@Templated` views you do not need to use `TranslationService` and handle setting i18n text programmatically. See [here](http://docs.jboss.org/errai/latest/errai/reference/html_single/#_html_template_translation). The combination if `data-i18n-prefix` and `data-i18n-key` _should_ be enough for Errai to handle _static_ translations automatically. If you need _dynamic_ translation (e.g. "There was an error processing file {0}" where "{0}" needs to be substituted you still need to follow the programmatic approach.

**Comment by gitgabrio** [comment]
@manstis 

> @gitgabrio You will always need _properties_ files to hold the i18n text; however with HTML `@Templated` views you do not need to use `TranslationService` and handle setting i18n text programmatically. See [here](http://docs.jboss.org/errai/latest/errai/reference/html_single/#_html_template_translation). The combination if `data-i18n-prefix` and `data-i18n-key` _should_ be enough for Errai to handle _static_ translations automatically. If you need _dynamic_ translation (e.g. "There was an error processing file {0}" where "{0}" needs to be substituted you still need to follow the programmatic approach.

_Thanks! _

File: kie-wb-common-widgets/kie-wb-common-ui/src/test/java/org/kie/workbench/common/widgets/client/assets/dropdown/SubmarineKieAssetsDropdownTest.java [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    private IsSubmarine isSubmarine;
+
+    @Mock
+    private Consumer<List<KieAssetsDropdownItem>> kieAssetsConsumer;
+
+    @Mock
+    private SubmarineKieAssetsDropdownView viewlocalMock;
+
+    @Mock
+    private KieAssetsDropdown dropdownLocal;
```
[the remark this answers is not in the document]
Comment (gitgabrio) [reply kie-wb-common-widgets/kie-wb-common-ui/src/test/java/org/kie/workbench/common/widgets/client/assets/dropdown/SubmarineKieAssetsDropdownTest.java]: @jomarko 

_Look above comment_

File: kie-wb-common-widgets/kie-wb-common-ui/src/test/java/org/kie/workbench/common/widgets/client/assets/dropdown/AbstractKieAssetsDropdownTest.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+    protected List<KieAssetsDropdownItem> assetList = IntStream.range(0, 3)
+            .mapToObj(i -> new KieAssetsDropdownItem("File_" + i + ".txt", "", DEFAULT_VALUE, new HashMap<>()))
+            .collect(Collectors.toList());
+
+    @Mock
+    private KieAssetsDropdownView viewlocalMock;
+
+    @Mock
+    private KieAssetsDropdown dropdownLocal;
```
[the remark this answers is not in the document]
Comment (gitgabrio) [reply kie-wb-common-widgets/kie-wb-common-ui/src/test/java/org/kie/workbench/common/widgets/client/assets/dropdown/AbstractKieAssetsDropdownTest.java]: @jomarko 

_Good catch, thx!_

File: kie-wb-common-widgets/kie-wb-common-ui/src/main/java/org/kie/workbench/common/widgets/client/assets/dropdown/AbstractKieAssetsDropdown.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        return this::assetListConsumerMethod;
+    }
+
+    protected void assetListConsumerMethod(List<KieAssetsDropdownItem> assetList) {
+        assetList.forEach(this::addValue);
+        view.refreshSelectPicker();
+        view.initialize();
+    }
+
+    protected List<KieAssetsDropdownItem> getKieAssets() {
```
[the remark this answers is not in the document]
Comment (gitgabrio) [reply kie-wb-common-widgets/kie-wb-common-ui/src/main/java/org/kie/workbench/common/widgets/client/assets/dropdown/AbstractKieAssetsDropdown.java]: @jomarko 

_Uhm, I've removed but I'm almost sure sooner or later some concrete class would need to expose them publicly, anyway...._

File: kie-wb-common-widgets/kie-wb-common-ui/src/main/java/org/kie/workbench/common/widgets/client/assets/dropdown/AbstractKieAssetsDropdown.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    public void addValue(final KieAssetsDropdownItem kieAsset) {
+        kieAssets.add(kieAsset);
+        view.addValue(kieAsset);
+    }
+
+    protected Consumer<List<KieAssetsDropdownItem>> getAssetListConsumer() {
+        return this::assetListConsumerMethod;
+    }
+
+    protected void assetListConsumerMethod(List<KieAssetsDropdownItem> assetList) {
```
[the remark this answers is not in the document]
Comment (gitgabrio) [reply kie-wb-common-widgets/kie-wb-common-ui/src/main/java/org/kie/workbench/common/widgets/client/assets/dropdown/AbstractKieAssetsDropdown.java]: @jomarko 

_Done_

**Comment by gitgabrio** [comment]
@manstis
Hi! Could you please merge this, squashing commits? Thanks!

**Pull request merged**